src/jit/SimdRiscvInl.h

```diff
@@ -0,0 +1,1251 @@
+{
+    rd = sljit_get_register_index(SLJIT_GP_REGISTER, rd);
+    rs1 = sljit_get_register_index(SLJIT_GP_REGISTER, rs1);
+    uint32_t opcode = 0x73 | ((uint32_t)rd << 7) | ((uint32_t)rs1 << 15) | (0x1 << 12) | (csr << 20);
+    sljit_emit_op_custom(compiler, &opcode, sizeof(uint32_t));
+}
+
+static void simdEmitAbs(sljit_compiler* compiler, sljit_s32 type, sljit_s32 rd, sljit_s32 rn)
+{
+    sljit_s32 mask = SLJIT_VR0;
+    sljit_s32 tmp = rd == rn ? SLJIT_TMP_DEST_VREG : rd;
+
+    simdEmitTypedOp(compiler, type, SimdOp::vmv_vv, tmp, 0, rn, SimdOp::rn_imm);
+    simdEmitOp(compiler, SimdOp::vmslt_vx, mask, rn, TMP_ZERO);
+    simdEmitOp(compiler, SimdOp::vrsub_vi ^ SimdOp::vm, tmp, rn, 0, SimdOp::rm_imm);
+
+    if (rd == rn) {
+        simdEmitOp(compiler, SimdOp::vmv_vv, rd, 0, tmp);
+    }
+}
+
+static void simdEmitAllTrue(sljit_compiler* compiler, sljit_s32 type, sljit_s32 rd, sljit_s32 rn)
+{
+    sljit_s32 tmp = SLJIT_TMP_DEST_VREG;
+    simdEmitTypedOp(compiler, type, SimdOp::vmv_vi, tmp, 0, (0x1F), SimdOp::rm_imm);
+    simdEmitOp(compiler, SimdOp::vredminu_vs, tmp, rn, tmp);
+    simdEmitOp(compiler, SimdOp::vmv_xs, rd, 0, tmp, SimdOp::rn_imm | SimdOp::rd_gpreg);
+    struct sljit_jump* notAllTrue = sljit_emit_cmp(compiler, SLJIT_EQUAL, rd, 0, SLJIT_IMM, 0);
+    sljit_emit_op1(compiler, SLJIT_MOV, rd, 0, SLJIT_IMM, 1);
+    sljit_set_label(notAllTrue, sljit_emit_label(compiler));
+}
+
+static void simdEmitAnyTrue(sljit_compiler* compiler, sljit_s32 type, sljit_s32 rd, sljit_s32 rn)
+{
+    sljit_s32 tmp = SLJIT_TMP_DEST_VREG;
+    simdEmitTypedOp(compiler, type, SimdOp::vmv_sx, tmp, 0, TMP_ZERO, SimdOp::rm_gpreg);
+    simdEmitOp(compiler, SimdOp::vredmaxu_vs, tmp, rn, tmp);
+    simdEmitOp(compiler, SimdOp::vmv_xs, rd, 0, tmp, SimdOp::rn_imm | SimdOp::rd_gpreg);
+    struct sljit_jump* notAnyTrue = sljit_emit_cmp(compiler, SLJIT_EQUAL, rd, 0, SLJIT_IMM, 0);
+    sljit_emit_op1(compiler, SLJIT_MOV, rd, 0, SLJIT_IMM, 1);
+    sljit_set_label(notAnyTrue, sljit_emit_label(compiler));
+}
+
+static void simdEmitAvgr(sljit_compiler* compiler, sljit_s32 type, sljit_s32 rd, sljit_s32 rn, sljit_s32 rm)
+{
+    sljit_s32 gptmp = SLJIT_TMP_DEST_REG;
+    simdEmitCSRRWI(compiler, gptmp, 0x00A, 0);
+    simdEmitTypedOp(compiler, type, SimdOp::vaaddu_vv, rd, rn, rm);
+    simdEmitCSRRW(compiler, gptmp, 0x00A, gptmp);
+}
+
+static void simdEmitCompare(sljit_compiler* compiler, sljit_s32 type, uint32_t opcode, sljit_s32 rd, sljit_s32 rn, sljit_s32 rm, bool reverseMask = false)
+{
+    sljit_s32 mask = SLJIT_VR0;
+    simdEmitTypedOp(compiler, type, opcode, mask, rn, rm);
+    sljit_s32 tmp = SLJIT_TMP_DEST_VREG;
+    simdEmitOp(compiler, SimdOp::vmv_vi, tmp, 0, reverseMask ? (0x1F) : 0, SimdOp::rm_imm | SimdOp::rn_imm);
+    simdEmitOp(compiler, SimdOp::vmerge_vi, rd, tmp, reverseMask ? 0 : (0x1F), SimdOp::rm_imm);
+}
+
+static void simdEmitExtend(sljit_compiler* compiler, sljit_s32 type, bool low, bool s, sljit_s32 rd, sljit_s32 rn)
```

Comment:
s = signed? Could have a longer name. What is low?

Reply:
`low` is `true` if the operator is `ExtendLow`, and `false` if the operator is `ExtendHigh`